core/trino-main/src/main/java/io/trino/operator/DeleteAndInsertMergeProcessor.java

```diff
@@ -0,0 +1,289 @@
+     */
+    @Override
+    public Page transformPage(Page inputPage)
+    {
+        requireNonNull(inputPage, "inputPage is null");
+        int inputChannelCount = inputPage.getChannelCount();
+        if (inputChannelCount < 2 + redistributionColumnCount) {
+            throw new IllegalArgumentException(format("inputPage channelCount (%s) should be >= 2 + partition columns size (%s)", inputChannelCount, redistributionColumnCount));
+        }
+
+        int originalPositionCount = inputPage.getPositionCount();
+        if (originalPositionCount <= 0) {
+            throw new IllegalArgumentException("originalPositionCount should be > 0, but is " + originalPositionCount);
+        }
+        Block mergeCaseBlock = inputPage.getBlock(redistributionColumnCount + 1);
+        List<Block> mergeCaseBlocks = mergeCaseBlock.getChildren();
+        int mergeBlocksSize = mergeCaseBlocks.size();
+        Block rowMatchedBlock = mergeCaseBlocks.get(mergeBlocksSize - 2);
+        Block operationChannelBlock = mergeCaseBlocks.get(mergeBlocksSize - 1);
+        duplicateRowFinder.checkForDuplicateTargetRows(inputPage, operationChannelBlock);
+
+        int updatePositions = 0;
+        int insertPositions = 0;
+        int deletePositions = 0;
+        for (int position = 0; position < originalPositionCount; position++) {
+            int operation = (int) INTEGER.getLong(operationChannelBlock, position);
+            switch (operation) {
+                case DEFAULT_CASE_OPERATION_NUMBER:
+                    break;
+                case INSERT_OPERATION_NUMBER:
+                    insertPositions++;
+                    break;
+                case DELETE_OPERATION_NUMBER:
+                    deletePositions++;
+                    break;
+                case UPDATE_OPERATION_NUMBER:
+                    updatePositions++;
+                    break;
+                default:
+                    throw new IllegalArgumentException("Unknown operator number " + operation);
+            }
+        }
+
+        int totalPositions = insertPositions + deletePositions + 2 * updatePositions;
+        Block rowIdBlock = inputPage.getBlock(redistributionColumnCount);
+
+        ImmutableList.Builder<Type> pageTypesBuilder = ImmutableList.builder();
+        pageTypesBuilder.addAll(dataColumnTypes);
+        pageTypesBuilder.add(INTEGER);
+        List<Type> pageTypes = pageTypesBuilder.build();
+        PageBuilder pageBuilder = new PageBuilder(totalPositions, pageTypes);
+        int resultPosition = 0;
+        int rowIdPosition = 0;
+        int[] rowIdPositions = new int[totalPositions];
+        for (int position = 0; position < originalPositionCount; position++) {
+            long operation = INTEGER.getLong(operationChannelBlock, position);
+            if (operation != DEFAULT_CASE_OPERATION_NUMBER) {
+                if (operation == DELETE_OPERATION_NUMBER || operation == UPDATE_OPERATION_NUMBER) {
```

Comment:
Why DELETE and UPDATE cases are handled the same way here?
Explain in a code comment.

Reply:
Both result in a delete row being written.  I'll add a comment to that effect.

Reply:
> This would affect SPI, so it's better to do it first iteration.

It has been done.